nixos/modules/services/monitoring/certspotter.nix

```diff
@@ -0,0 +1,126 @@
+{ config
+, lib
+, pkgs
+, ... }:
+
+let
+  cfg = config.services.certspotter;
+
+  configDir = pkgs.linkFarm "certspotter-config"
+    (lib.toList {
+      name = "watchlist";
+      path = pkgs.writeText "cerspotter-watchlist" (builtins.concatStringsSep "\n" cfg.watchlist);
+    }
+    ++ lib.optional (cfg.emailRecipients != [ ]) {
+      name = "email_recipients";
+      path = pkgs.writeText "cerspotter-email_recipients" (builtins.concatStringsSep "\n" cfg.emailRecipients);
+    }
+    ++ lib.optional (cfg.hooks != [ ]) {
+      name = "hooks.d";
+      path = pkgs.linkFarm "certspotter-hooks" (lib.imap1 (i: path: {
+        inherit path;
+        name = "hook${toString i}";
+      }) cfg.hooks);
+    });
+in
+{
+  options.services.certspotter = {
+    enable = lib.mkEnableOption "Cert Spotter, a Certificate Transparency log monitor";
+
+    sendmailPath = lib.mkOption {
+      type = lib.types.path;
+      description = ''
+        Path to the `sendmail` binary. By default, the local sendmail wrapper is used
+        (see `config.services.mail.sendmailSetuidWrapper`).
+      '';
+      example = lib.literalExpression ''"''${pkgs.system-sendmail}/bin/sendmail"'';
+    };
+
+    watchlist = lib.mkOption {
+      type = with lib.types; listOf str;
+      description = "Domain names to watch. To monitor a domain with all subdomains, prefix its name with `.` (e.g. `.example.org`).";
+      default = [ ];
+      example = [ ".example.org" "another.example.com" ];
+    };
+
+    emailRecipients = lib.mkOption {
+      type = with lib.types; listOf str;
+      description = "A list of email addresses to send certificate updates to.";
+      default = [ ];
+    };
+
+    hooks = lib.mkOption {
+      type = with lib.types; listOf path;
+      description = ''
+        Scripts to run upon the detection of a new certificate. See `man 8 certspotter-script` or [the GitHub page](https://github.com/SSLMate/certspotter/blob/master/man/certspotter-script.md) for more info.
+      '';
+      default = [];
+      example = lib.literalExpression ''
+        [
+          (pkgs.writeShellScript "certspotter-hook" '''
+            echo "Event summary: $SUMMARY."
+          ''')
+        ]
+      '';
+    };
+
+    extraFlags = lib.mkOption {
+      type = with lib.types; listOf str;
+      description = "Extra command-line arguments to pass to Cert Spotter";
+      example = [ "-start_at_end" ];
+      default = [ ];
```

Comment:
I'd suggest passing `-start_at_end` by default. The reason for that is that adding entries to the watchlist causes certspotter to watch those from the current point in time anyway, so not having start_at_end gives a false sense of security for people who build the watchlist in a way that allows it to grow.

Reply:
my concern with this is downtimes - if certspotter has been down for some time (e.g. because of a power outage or server reboot), a bad actor could issue a certificate within that timeframe. `start_at_end` means everything during the downtime will be ignored

Reply:
AFAIU (from certspotter documentation, will confirm with its code in the evening), `start_at_end` only matters when initializing for the first time (i.e. when the state file is not present).

Reply:
Not exactly, it matters when initializing *each CT log*. CT logs are added every year, so this can still make CT spotter miss some logs.

Reply:
I've checked in sources: https://github.com/SSLMate/certspotter/blob/master/monitor/monitor.go#L104

It is also indeed per-log, and that makes for a better argument to keep it on by default: otherwise you'll incur the large amount of traffic at apparently random points in time when a new log gets added, for a very small benefit (it only surfaces certificates that were added to the new log before the log's addition and weren't added earlier to any previously-listed logs and will not get added to them momentarily).

In any case, would you mind making start_at_end a separate option instead of requiring it to be specified via generic commandline flags option? That would make for easier overriding in both directions.

Reply:
I've made the module add `-start_at_end` if no logs have been checked yet. This should work fine in most situations.